File: src/dmocr/web/system.py
"""Data for the System view.

The Review board answers "what does a Risk Manager need to do about this case?". This
answers a different question: **"what is the system actually doing, and what is it not
doing yet?"**

That distinction matters because the honest answers are frequently negative — 3 of 21
requirements are blocked on documents we do not hold, no rule is approved, only one
verification source is automatable, and there are no accuracy numbers on real documents.
Those facts live in YAML and Markdown today, which means they are invisible while you
work. Surfacing them is the point.

Everything here is read-only and derived. Nothing in this module changes a case.
"""

from __future__ import annotations

import json
import logging
import re
from pathlib import Path

import yaml

log = logging.getLogger(__name__)

REPO_ROOT = Path(__file__).resolve().parents[3]
REQUIREMENTS_PATH = REPO_ROOT / "docs/regulatory/requirements.yaml"
SOURCES_PATH = REPO_ROOT / "docs/regulatory/sources.yaml"
OPEN_ITEMS_PATH = REPO_ROOT / "docs/OPEN-ITEMS.md"
EVAL_REPORT_PATH = REPO_ROOT / "eval-output/evaluation.json"


def _load_yaml(path: Path) -> dict:
    try:
        return yaml.safe_load(path.read_text(encoding="utf-8")) or {}
    except Exception as exc:
        log.warning("could not read %s: %s", path, exc)
        return {}


# =====================================================================================
# Rules
# =====================================================================================


def rules_view(session) -> dict:
    """Every rule, with what it did on the current case.

    A rule that produced nothing is as interesting as one that fired - "why did this not
    check anything?" is the question you actually have while tuning.
    """
    if session.rule_set is None:
        return {"version": None, "rules": [], "note": "No rule set loaded."}

    findings = {f.rule_id: f for f in session.findings}
    requirements = {
        r["id"]: r for r in (_load_yaml(REQUIREMENTS_PATH).get("requirements") or [])
    }

    rules = []
    for rule in session.rule_set.rules:
        finding = findings.get(rule.rule_id)
        cites = [
            {
                "id": c,
                "location": requirements.get(c, {}).get("location"),
                "quote": (requirements.get(c, {}).get("quote") or "")[:300],
                "resolves": c in requirements,
            }
            for c in rule.citations
        ]
        rules.append({
            "rule_id": rule.rule_id,
            "title": rule.title,
            "category": rule.category,
            "severity": rule.severity.value,
            "determinacy": rule.determinacy.value,
            "status": rule.status.value,
            "enforceable": rule.is_enforceable,
            "check": rule.check,
            "params": rule.params,
            # An empty citation list is meaningful: a BUSINESS rule, not a regulatory
            # checkpoint. The UI must never imply backing a rule does not have.
            "regulatory": rule.is_regulatory,
            "citations": cites,
            "applicability": {
                k: v for k, v in
                (rule.applicability.model_dump() if rule.applicability else {}).items()
                if v
            },
            "recommended_action": rule.recommended_action,
            "outcome": None if finding is None else {
                "determination": finding.determination.value,
                "disposition": finding.disposition.value,
                "message": finding.message,
                "advisory": finding.advisory_only,
                "evidence": finding.evidence.note,
            },
        })

    approved = sum(1 for r in session.rule_set.rules if r.is_enforceable)
    return {
        "version": session.rule_set.version,
        "total": len(rules),
        "approved": approved,
        "regulatory": sum(1 for r in rules if r["regulatory"]),
        "business": sum(1 for r in rules if not r["regulatory"]),
        "rules": rules,
        "note": (
            "No rule is APPROVED, so nothing is enforced. The board evaluates in DRY_RUN "
            "and labels every finding advisory. Rules ship disabled until legal sign-off."
            if approved == 0 else None
        ),
    }


# =====================================================================================
# Regulatory knowledge base
# =====================================================================================

#: A requirement may only become a rule when its source is verified.
RULE_READY_STATUS = "PRIMARY_VERIFIED"


def regulatory_view() -> dict:
    """The 21 requirements, with what blocks the ones that are blocked."""
    reqs_doc = _load_yaml(REQUIREMENTS_PATH)
    sources_doc = _load_yaml(SOURCES_PATH)
    instruments = {i["id"]: i for i in (sources_doc.get("instruments") or [])}

    requirements = []
    for r in reqs_doc.get("requirements") or []:
        inst = instruments.get(r.get("source"), {})
        verification = inst.get("verification_status", "UNKNOWN")
        needs_review = r.get("status") == "REQUIRES_LEGAL_REVIEW"
        blocked_because = None
        if verification != RULE_READY_STATUS:
            blocked_because = (
                f"Source {inst.get('id', r.get('source'))} is {verification}, not "
                f"{RULE_READY_STATUS}. It must be confirmed against an authoritative "
                f"copy before this can become a rule."
            )
        elif needs_review:
            blocked_because = (
                "Flagged REQUIRES_LEGAL_REVIEW - the requirement is ambiguous and must "
                "not be auto-enabled."
            )
        requirements.append({
            "id": r["id"],
            "source": r.get("source"),
            "instrument": inst.get("title"),
            "location": r.get("location"),
            "quote": r.get("quote"),
            "feasibility": r.get("feasibility"),
            "obligation_kind": r.get("obligation_kind"),
            "priority": r.get("priority"),
            "notes": r.get("notes"),
            "verification_status": verification,
            "rule_ready": blocked_because is None,
            "blocked_because": blocked_because,
        })

    negatives = [
        {"id": n["id"], "source": n.get("source"),
         "searched_for": n.get("searched_for"),
         "conclusion": n.get("conclusion"), "consequence": n.get("consequence")}
        for n in (reqs_doc.get("negative_findings") or [])
    ]

    return {
        "requirements": requirements,
        "total": len(requirements),
        "rule_ready": sum(1 for r in requirements if r["rule_ready"]),
        "blocked": sum(1 for r in requirements if not r["rule_ready"]),
        # What an instrument does NOT say is a finding too, and it is the kind that gets
        # silently re-derived - or invented - if it is not written down.
        "negative_findings": negatives,
        "instruments": [
            {"id": i["id"], "title": i.get("title"),
             "status": i.get("verification_status"),
             "reference": i.get("reference")}
            for i in instruments.values()
        ],
        "provenance": sources_doc.get("local_copy_provenance") or [],
    }


# =====================================================================================
# Pipeline trace
# =====================================================================================


def trace_view(session) -> dict:
    """Stage-by-stage detail for each uploaded document."""
    from .service import box_label

    documents = []
    for ctx in session.documents.values():
        extraction = ctx.extraction
        documents.append({
            "document_id": ctx.document_id,
            "box": box_label(ctx.box_key),
            "filename": ctx.filename,
            "status": ctx.status.value,
            "size_bytes": len(ctx.data),
            "quality": ctx.quality_metrics,
            "reading": ctx.ocr_stats,
            "classification": ctx.classification_detail,
            "extraction": None if extraction is None else {
                "fields": extraction.extracted_count,
                "missing_required": list(extraction.missing_required),
                # Values the model produced but could not point at on the page. Every one
                # of these is a hallucination that ADR-0004 caught.
                "discarded_ungrounded": list(extraction.rejected_ungrounded),
                "notes": list(extraction.notes),
                "by_field": [
                    {"name": f.field_name, "attribute": f.attribute,
                     "confidence": f.confidence.value, "page": f.provenance.page,
                     "source": f.provenance.kind,
                     "notes": list(f.notes)}
                    for f in extraction.fields
                ],
            },
            "corrections": sorted(ctx.corrections),
            "stages": [
                {"key": s.key, "label": s.label, "status": s.status.value,
                 "detail": s.detail}
                for s in ctx.stages
            ],
        })

    prop = session.case.properties[0] if session.case.properties else None
    claims = []
    if prop is not None:
        for attribute, claim_set in sorted(prop.claim_sets.items()):
            resolution = prop.resolve(attribute)
            claims.append({
                "attribute": attribute,
                "claims": len(claim_set.claims),
                "determination": resolution.determination.value,
                "rationale": resolution.rationale,
                "confidence": resolution.confidence.value,
            })

    return {
        "documents": documents,
        "case_id": session.case.case_id,
        "claims": claims,
        "parties": [
            {"roles": sorted(p.roles), "variants": p.name_variants}
            for p in session.case.parties
        ],
        "notes": list(session.case_notes),
        "context": (
            None if session.case.processing_context is None else {
                "pipeline_version": session.case.processing_context.pipeline_version,
                "rule_set_version": session.case.processing_context.rule_set_version,
                "models": session.case.processing_context.model_versions,
                "regulatory_as_of": session.case.processing_context.regulatory_as_of.isoformat(),
            }
        ),
    }


# =====================================================================================
# Verification
# =====================================================================================


def verification_view(session) -> dict:
    run = session.verification
    if run is None:
        return {"planned": False,
                "note": "No property on the case yet, so nothing has been planned."}

    return {
        "planned": True,
        "summary": run.summary(),
        "plan": [
            {
                "source_id": item.source_id,
                "authority": item.source.authority,
                "tier": item.tier.value,
                "tier_confidence": item.source.tier_confidence,
                "execution": item.execution.value,
                "attributes": item.attributes,
                # Exactly what would be sent. An external lookup is an outbound
                # disclosure of customer data, so the minimisation decision is shown.
                "lookup_keys": item.lookup_keys,
                "ambiguous_keys": item.ambiguous_keys,
                "reason": item.reason,
                "blocked_on": list(item.source.blocked_on),
            }
            for item in run.plan.items
        ],
        "tasks": [
            {"task_id": t.task_id, "source_id": t.source_id, "authority": t.authority,
             "tier": t.tier.value, "status": t.status.value,
             "lookup_keys": t.lookup_keys, "attributes": t.attributes,
             "instruction": t.render_instruction()}
            for t in run.tasks
        ],
        "results": [
            {"source_id": r.source_id, "attribute": r.attribute,
             "status": r.status.value, "detail": r.detail,
             "internal": r.internal_value, "external": r.external_value}
            for r in run.results
        ],
        "notes": list(run.notes),
    }


# =====================================================================================
# Evaluation
# =====================================================================================


def evaluation_view() -> dict:
    """The last harness run, if one has been done."""
    if not EVAL_REPORT_PATH.is_file():
        return {
            "available": False,
            "note": ("No evaluation has been run in this workspace. Run it from this "
                     "page, or with tools/evaluate.py."),
        }
    try:
        report = json.loads(EVAL_REPORT_PATH.read_text(encoding="utf-8"))
    except Exception as exc:
        return {"available": False, "note": f"Could not read the report: {exc}"}

    return {
        "available": True,
        "started_at": report.get("started_at"),
        "coverage": report.get("coverage", {}),
        "classification": report.get("classification", {}),
        "ocr": report.get("ocr", {}),
        "extraction": report.get("extraction", {}).get("overall", {}),
        "by_field": report.get("extraction", {}).get("by_field", {}),
        "findings": report.get("findings", {}),
        "dangerous_errors": report.get("dangerous_errors", []),
        "notes": report.get("notes", []),
        # Reading a rate over a handful of synthetic documents as an accuracy figure is
        # the mistake this warning exists to prevent.
        "caveat": (
            "Measured on synthetic fixtures. The values are the ones the generator wrote, "
            "so this demonstrates plumbing, not competence on real documents."
        ),
    }


# =====================================================================================
# Open items
# =====================================================================================

_ROW = re.compile(r"^\|\s*(?P<num>~*\d+~*)\s*\|\s*(?P<item>.+?)\s*\|(?P<rest>.*)\|\s*$")


def open_items_view() -> dict:
    """The tracked deferrals, parsed out of OPEN-ITEMS.md.

    Read from the file rather than duplicated, so the board cannot drift from the record.
    """
    if not OPEN_ITEMS_PATH.is_file():
        return {"items": [], "note": "OPEN-ITEMS.md not found."}

    section = ""
    items = []
    for line in OPEN_ITEMS_PATH.read_text(encoding="utf-8").splitlines():
        if line.startswith("## "):
            section = line[3:].strip()
            continue
        m = _ROW.match(line)
        if not m:
            continue
        num = m.group("num")
        closed = num.startswith("~")
        detail = " ".join(
            c.strip() for c in m.group("rest").split("|") if c.strip())
        items.append({
            "number": num.strip("~"),
            "section": section,
            "item": m.group("item").replace("**", "").replace("~~", "").strip(),
            "detail": detail,
            "closed": closed,
        })

    open_count = sum(1 for i in items if not i["closed"])
    return {
        "items": items,
        "total": len(items),
        "open": open_count,
        "closed": len(items) - open_count,
        "by_section": {
            s: sum(1 for i in items if i["section"] == s and not i["closed"])
            for s in dict.fromkeys(i["section"] for i in items)
        },
    }

File: src/dmocr/web/static/system.js
"use strict";

// System view. Read-only: nothing here changes a case, except the explicit
// "run evaluation" button.

const PANELS = {
  rules: ["/api/system/rules", renderRules],
  regulatory: ["/api/system/regulatory", renderRegulatory],
  trace: ["/api/system/trace", renderTrace],
  verification: ["/api/system/verification", renderVerification],
  evaluation: ["/api/system/evaluation", renderEvaluation],
  "open-items": ["/api/system/open-items", renderOpenItems],
};

let current = "rules";
let evalPoll = null;

async function load(name) {
  const [url, render] = PANELS[name];
  const host = document.getElementById("panel-" + name);
  if (!host.innerHTML) host.innerHTML = `<p class="empty">Loading…</p>`;
  try {
    host.innerHTML = render(await (await fetch(url)).json());
  } catch (e) {
    host.innerHTML = `<p class="empty">Could not load: ${esc(e.message)}</p>`;
  }
}

function show(name) {
  current = name;
  Object.keys(PANELS).forEach((k) => {
    document.getElementById("panel-" + k).hidden = k !== name;
  });
  document.querySelectorAll(".panelnav .tab").forEach((b) =>
    b.classList.toggle("on", b.dataset.panel === name));
  load(name);
}

// ---------------------------------------------------------------- rules

function renderRules(d) {
  if (!d.rules.length) return `<p class="empty">${esc(d.note || "No rules.")}</p>`;

  let html = `<div class="summary">
    <div class="stat"><b>${d.total}</b>rules</div>
    <div class="stat ${d.approved ? "cleared" : "review"}"><b>${d.approved}</b>approved</div>
    <div class="stat"><b>${d.regulatory}</b>regulatory</div>
    <div class="stat"><b>${d.business}</b>business</div>
  </div>`;
  if (d.note) html += `<p class="callout">${esc(d.note)}</p>`;

  html += d.rules.map((r) => {
    const o = r.outcome;
    const disp = o ? o.disposition : "—";
    return `<details class="row">
      <summary>
        <span class="mono">${esc(r.rule_id)}</span>
        <span class="rtitle">${esc(r.title)}</span>
        <span class="tag">${r.severity}</span>
        <span class="tag ${r.regulatory ? "reg" : ""}">${
          r.regulatory ? "regulatory" : "business rule"}</span>
        <span class="tag ${r.status === "APPROVED" ? "" : "warn"}">${r.status}</span>
        ${o ? `<span class="tag ${o.disposition}">${esc(disp)}</span>` : ""}
      </summary>
      <div class="rowbody">
        <div class="kv"><span>Checks</span><code>${esc(r.check)}</code></div>
        ${Object.keys(r.params || {}).length
          ? `<div class="kv"><span>Parameters</span><code>${esc(JSON.stringify(r.params))}</code></div>` : ""}
        ${Object.keys(r.applicability || {}).length
          ? `<div class="kv"><span>Applies when</span><code>${esc(JSON.stringify(r.applicability))}</code></div>` : ""}
        <div class="kv"><span>Determinacy</span><span>${esc(r.determinacy)}</span></div>
        ${o ? `<div class="kv"><span>On this case</span>
            <span><strong>${esc(o.determination)}</strong> &rarr; ${esc(o.disposition)}
            ${o.advisory ? `<span class="tag">advisory</span>` : ""}<br>
            <span class="muted">${esc(o.message)}</span></span></div>`
           : `<div class="kv"><span>On this case</span><span class="muted">not evaluated — no documents yet</span></div>`}
        ${r.citations.length
          ? `<div class="kv"><span>Cites</span><span>${r.citations.map((c) =>
              `<div class="cite ${c.resolves ? "" : "bad"}">
                 <code>${esc(c.id)}</code> ${c.location ? esc(c.location) : ""}
                 ${c.quote ? `<blockquote>${esc(c.quote)}…</blockquote>` : ""}
               </div>`).join("")}</span></div>`
          : `<div class="kv"><span>Cites</span><span class="muted">nothing — this is a
              business rule, not a regulatory checkpoint</span></div>`}
        ${r.recommended_action
          ? `<div class="kv"><span>Action</span><span>${esc(r.recommended_action)}</span></div>` : ""}
      </div></details>`;
  }).join("");
  return html;
}

// ---------------------------------------------------------------- regulatory

function renderRegulatory(d) {
  let html = `<div class="summary">
    <div class="stat"><b>${d.total}</b>requirements</div>
    <div class="stat cleared"><b>${d.rule_ready}</b>rule-ready</div>
    <div class="stat review"><b>${d.blocked}</b>blocked</div>
    <div class="stat"><b>${d.instruments.length}</b>instruments</div>
  </div>
  <p class="callout">A requirement may only become a rule when its source has been read
  from an authoritative copy and it is not flagged for legal review.</p>`;

  html += d.requirements.map((r) => `
    <details class="row ${r.rule_ready ? "" : "blocked"}">
      <summary>
        <span class="mono">${esc(r.id)}</span>
        <span class="tag ${r.rule_ready ? "" : "warn"}">${
          r.rule_ready ? "rule-ready" : "blocked"}</span>
        <span class="tag">${esc(r.feasibility || "")}</span>
        <span class="muted">${esc(r.location || "")}</span>
      </summary>
      <div class="rowbody">
        ${r.blocked_because ? `<p class="callout warn">${esc(r.blocked_because)}</p>` : ""}
        <div class="kv"><span>Instrument</span><span>${esc(r.instrument || r.source)}</span></div>
        <blockquote>${esc(r.quote || "")}</blockquote>
        ${r.notes ? `<div class="kv"><span>Notes</span><span class="muted">${esc(r.notes)}</span></div>` : ""}
      </div></details>`).join("");

  if (d.negative_findings.length) {
    html += `<h3>Negative findings</h3>
      <p class="muted small">What an instrument does NOT say is a finding too — and the
      kind that gets silently re-derived, or invented, if it is not written down.</p>`;
    html += d.negative_findings.map((n) => `
      <details class="row"><summary><span class="mono">${esc(n.id)}</span>
        <span class="muted">${esc((n.conclusion || "").slice(0, 90))}…</span></summary>
        <div class="rowbody">
          <div class="kv"><span>Searched for</span><code>${esc(n.searched_for || "")}</code></div>
          <div class="kv"><span>Conclusion</span><span>${esc(n.conclusion || "")}</span></div>
          <div class="kv"><span>Consequence</span><span>${esc(n.consequence || "")}</span></div>
        </div></details>`).join("");
  }
  return html;
}

// ---------------------------------------------------------------- trace

function renderTrace(d) {
  if (!d.documents.length)
    return `<p class="empty">No documents yet. Upload one on the Review tab.</p>`;

  let html = "";
  if (d.context) {
    html += `<p class="callout">Pinned for reproducibility —
      pipeline ${esc(d.context.pipeline_version)}, rules ${esc(d.context.rule_set_version)},
      OCR ${esc(d.context.models.ocr || "n/a")}, regulation as at
      ${esc(d.context.regulatory_as_of)}.</p>`;
  }

  html += d.documents.map((doc) => {
    const q = doc.quality || {}, r = doc.reading || {}, c = doc.classification || {},
          x = doc.extraction;
    return `<details class="row" open>
      <summary><span class="rtitle">${esc(doc.box)}</span>
        <span class="muted">${esc(doc.filename)}</span>
        <span class="tag ${doc.status}">${esc(doc.status)}</span></summary>
      <div class="rowbody">
        <div class="kv"><span>File</span><span>${(doc.size_bytes/1024).toFixed(0)} KB ·
          sha ${esc(q.sha256 || "")} · safety ${esc(q.safety || "")}</span></div>
        <div class="kv"><span>Quality</span><span>${esc(q.verdict || "")} ·
          ${q.page_count || "?"} page(s) · text layer: ${q.has_text_layer ? "yes" : "no"}
          ${(q.notes || []).length ? `<br><span class="muted">${esc((q.notes||[]).join("; "))}</span>` : ""}</span></div>
        <div class="kv"><span>Reading</span><span>${r.text_layer_pages ?? 0} from text layer,
          ${r.ocr_pages ?? 0} by OCR${r.mean_confidence != null
            ? ` · mean OCR confidence ${(r.mean_confidence*100).toFixed(1)}%` : ""}
          ${(r.failures||[]).length ? `<br><span class="warn">${esc(r.failures.join("; "))}</span>` : ""}</span></div>
        <div class="kv"><span>Classification</span><span>${esc(c.predicted || "—")}
          (${esc(c.confidence || "")}, score ${c.score ?? "—"})
          ${c.unknown_reason ? ` · reason ${esc(c.unknown_reason)}` : ""}
          ${c.scores ? `<br><span class="muted">all scores: ${esc(JSON.stringify(c.scores))}</span>` : ""}</span></div>
        ${(c.signals || []).length ? `<div class="kv"><span>Signals matched</span>
          <span class="muted small">${c.signals.map((s) =>
            `${esc(s.name)} p${s.page} “${esc(s.text)}” +${s.weight}`).join("<br>")}</span></div>` : ""}
        ${x ? `<div class="kv"><span>Extraction</span><span>${x.fields} field(s)
            ${x.missing_required.length
              ? `<br><span class="warn">missing required: ${esc(x.missing_required.join(", "))}</span>` : ""}
            ${x.discarded_ungrounded.length
              ? `<br><span class="warn">discarded (not locatable on the page):
                 ${esc(x.discarded_ungrounded.join("; "))}</span>` : ""}</span></div>
          <div class="kv"><span>Fields</span><span class="muted small">${x.by_field.map((f) =>
            `${esc(f.name)} → ${esc(f.attribute)} · p${f.page} · ${esc(f.confidence)} · via ${esc(f.source)}`
          ).join("<br>")}</span></div>` : ""}
        ${doc.corrections.length ? `<div class="kv"><span>Corrected</span>
          <span>${esc(doc.corrections.join(", "))}</span></div>` : ""}
      </div></details>`;
  }).join("");

  if (d.claims.length) {
    html += `<h3>Canonical claims</h3>
      <p class="muted small">Several sources may assert the same attribute. Disagreement is
      preserved, not resolved — that disagreement is the finding.</p>
      <table class="grid"><thead><tr><th>Attribute</th><th>Claims</th>
        <th>Resolves to</th><th>Why</th></tr></thead><tbody>` +
      d.claims.map((c) => `<tr><td><code>${esc(c.attribute)}</code></td>
        <td>${c.claims}</td>
        <td><span class="tag ${c.determination}">${esc(c.determination)}</span></td>
        <td class="muted">${esc(c.rationale)}</td></tr>`).join("") +
      `</tbody></table>`;
  }
  if (d.parties.length) {
    html += `<h3>Resolved parties</h3>` + d.parties.map((p) =>
      `<div class="kv"><span>${esc(p.roles.join(", "))}</span>
       <span>${esc(p.variants.join("  ·  "))}</span></div>`).join("");
  }
  return html;
}

// ---------------------------------------------------------------- verification

function renderVerification(d) {
  if (!d.planned) return `<p class="empty">${esc(d.note)}</p>`;
  const s = d.summary;

  let html = `<div class="summary">
    <div class="stat"><b>${s.results}</b>results</div>
    <div class="stat ${s.checks_performed ? "cleared" : "review"}"><b>${s.checks_performed}</b>answered</div>
    <div class="stat review"><b>${s.pending_manual}</b>pending</div>
    <div class="stat"><b>${s.source_unavailable}</b>unavailable</div>
  </div>
  <p class="callout">An unavailable source is <strong>never</strong> a compliance failure.
  Unanswered checks count toward case completeness, never toward pass or fail.</p>
  <table class="grid"><thead><tr><th>Source</th><th>Tier</th><th>How</th>
    <th>Would send</th><th>Why</th></tr></thead><tbody>` +
    d.plan.map((p) => `<tr>
      <td><code>${esc(p.source_id)}</code><br><span class="muted small">${esc(p.authority)}</span></td>
      <td>${esc(p.tier)}<br><span class="muted small">${esc(p.tier_confidence)}</span></td>
      <td><span class="tag ${p.execution === "AUTOMATED" ? "" : "warn"}">${esc(p.execution)}</span></td>
      <td><code>${esc(JSON.stringify(p.lookup_keys))}</code>
        ${p.ambiguous_keys.length ? `<br><span class="warn small">disputed, not used:
          ${esc(p.ambiguous_keys.join(", "))}</span>` : ""}</td>
      <td class="muted small">${esc(p.reason)}</td></tr>`).join("") +
    `</tbody></table>`;

  if (d.tasks.length) {
    html += `<h3>Operator tasks</h3>
      <p class="muted small">Only one registered source can be automated. These are the
      rest — the operator supplies access, the system supplies the comparison and audit.</p>`;
    html += d.tasks.map((t) => `<pre class="task">${esc(t.instruction)}</pre>`).join("");
  }
  if (d.notes.length)
    html += `<h3>Notes</h3><ul class="muted small">` +
      d.notes.map((n) => `<li>${esc(n)}</li>`).join("") + `</ul>`;
  return html;
}

// ---------------------------------------------------------------- evaluation

function renderEvaluation(d) {
  let html = `<div class="actions">
    <button id="run-eval" ${d.running ? "disabled" : ""}>
      ${d.running ? "Running…" : "Run evaluation"}</button>
    ${d.running ? `<span class="muted">This takes about a minute.</span>` : ""}
  </div>`;
  if (d.last_error) html += `<p class="callout warn">${esc(d.last_error)}</p>`;
  if (!d.available) return html + `<p class="empty">${esc(d.note)}</p>`;

  const x = d.extraction || {}, c = d.classification || {}, o = d.ocr || {};
  const pct = (v) => v == null ? "n/a" : (v * 100).toFixed(1) + "%";

  html += `<p class="callout warn">${esc(d.caveat)}</p>
    <div class="summary">
      <div class="stat"><b>${d.coverage.documents ?? 0}</b>documents</div>
      <div class="stat"><b>${d.coverage.labelled_fields ?? 0}</b>labelled fields</div>
      <div class="stat cleared"><b>${pct(x.precision)}</b>precision</div>
      <div class="stat cleared"><b>${pct(x.recall)}</b>recall</div>
      <div class="stat ${x.dangerous_error_rate ? "blocker" : "cleared"}">
        <b>${pct(x.dangerous_error_rate)}</b>dangerous errors</div>
    </div>
    <p class="muted small">“Dangerous” means a value that was produced and was wrong —
    the kind that reaches a reviewer as an answer rather than as a gap. A missing value is
    a safe failure and is counted separately.</p>
    <h3>Classification</h3>
    <div class="kv"><span>Accuracy on decided</span><span>${pct(c.accuracy_on_decided)}</span></div>
    <div class="kv"><span>Deferral rate</span><span>${pct(c.deferral_rate)}
      <span class="muted">— routed to a human, not counted as error</span></span></div>
    <h3>OCR</h3>
    <div class="kv"><span>CER / WER</span><span>${pct(o.cer_mean)} / ${pct(o.wer_mean)}
      <span class="muted">— a large gap means characters read but word boundaries lost</span></span></div>`;

  const fields = Object.entries(d.by_field || {});
  if (fields.length) {
    html += `<h3>By field</h3><table class="grid"><thead><tr><th>Field</th><th>Correct</th>
      <th>Near</th><th>Wrong</th><th>Missing</th><th>Spurious</th></tr></thead><tbody>` +
      fields.map(([n, v]) => `<tr><td><code>${esc(n)}</code></td><td>${v.correct}</td>
        <td>${v.near}</td><td class="${v.wrong ? "warn" : ""}">${v.wrong}</td>
        <td>${v.missing}</td><td class="${v.spurious ? "warn" : ""}">${v.spurious}</td></tr>`).join("") +
      `</tbody></table>`;
  }
  if ((d.notes || []).length)
    html += `<h3>Notes</h3><ul class="muted small">` +
      d.notes.map((n) => `<li>${esc(n)}</li>`).join("") + `</ul>`;
  return html;
}

// ---------------------------------------------------------------- open items

function renderOpenItems(d) {
  if (!d.items.length) return `<p class="empty">${esc(d.note || "None.")}</p>`;
  let html = `<div class="summary">
    <div class="stat review"><b>${d.open}</b>open</div>
    <div class="stat cleared"><b>${d.closed}</b>closed</div>
  </div>`;
  const sections = [...new Set(d.items.map((i) => i.section))];
  for (const s of sections) {
    html += `<h3>${esc(s)}</h3><table class="grid"><tbody>` +
      d.items.filter((i) => i.section === s).map((i) => `
        <tr class="${i.closed ? "done" : ""}">
          <td class="num">${esc(i.number)}</td>
          <td>${i.closed ? "<s>" : ""}${esc(i.item)}${i.closed ? "</s>" : ""}
            <div class="muted small">${esc(i.detail)}</div></td>
        </tr>`).join("") + `</tbody></table>`;
  }
  return html;
}

// ---------------------------------------------------------------- wiring

document.addEventListener("click", async (ev) => {
  const t = ev.target;
  if (t.dataset.panel) return show(t.dataset.panel);
  if (t.id === "refresh") return load(current);
  if (t.id === "run-eval") {
    t.disabled = true; t.textContent = "Running…";
    await fetch("/api/system/run-evaluation", { method: "POST" });
    if (!evalPoll) {
      evalPoll = setInterval(async () => {
        const d = await (await fetch("/api/system/evaluation")).json();
        if (!d.running) { clearInterval(evalPoll); evalPoll = null; load("evaluation"); }
      }, 2000);
    }
  }
});

function esc(s) {
  return String(s ?? "").replace(/[&<>"']/g,
    (c) => ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;" }[c]));
}

show("rules");
